Add daikin zone temperature control

barneyonline opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add zone temperature control to supported daikin systems based on [PR #93782](https://github.com/home-assistant/core/pull/93782)

<img width="472" alt="Screenshot 2024-07-09 at 4 09 25 PM" src="https://github.com/home-assistant/core/assets/38914183/ae07abde-0384-43cb-85c6-5d5e5ba50ae1">
<img width="829" alt="Screenshot 2024-07-09 at 4 10 06 PM" src="https://github.com/home-assistant/core/assets/38914183/ff2a95ac-7e10-4287-b8b7-5a52fb4f4177">


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue:
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33627

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

disforw commented:
Nice add! Are you assuming all Daikin units use Celsius as the native unit?

barneyonline commented:
> Nice add! Are you assuming all Daikin units use Celsius as the native unit?

Thanks! Yes, this aligns with the current integration. Can only speculate why, probably because it is the superior unit of measurement...

barneyonline committed:
PR 118527

For unit testing

frenck reviewed (changes requested):
Hi there @barneyonline 👋 

There seems to be a merge conflict on this PR. Could you take a look?

Thanks! 👍 

../Frenck

barneyonline committed:
Merge remote-tracking branch 'upstream/dev' into dev

barneyonline commented:
Thanks @frenck, conflict has been resolved

joostlek reviewed (changes requested):
Please fix the CI. Also, .coveragerc has been deleted from the project

barneyonline committed:
Merge branch 'home-assistant:dev' into dev

barneyonline committed:
Delete .coveragerc

barneyonline committed:
Merge branch 'home-assistant:dev' into dev

barneyonline commented:
CI fixed and .coveragerc removed

joostlek reviewed (commented):
Okay so wait, let me actually get this right, a zone is for example a different room and also has a setpoint and a state? Shouldn't it also be a climate entity? It could then be a new device with this climate entity so you can assign it to the right room. wdyt?

barneyonline committed:
fix ruff attempt1

barneyonline committed:
fix ruff attempt2

fredrike commented:
I would like to see a test case for the cases of all the different models that don't have zones.

barneyonline committed:
Test cases for zone temperature control

And hide temperature controls on daikin devices with zones but no zone temperature control

barneyonline commented:
@joostlek Any further thoughts on use of number entity here?
@fredrike Updated with unit tests for zone temperature control and updated number.py to better handle devices without support.

joostlek commented:
Btw, please rebase onto dev, last week Daikin got an update with a base entity you can use.

My main concern is that it ideally should be a different device so those devices can be connected to the right area. And the second concern is that I'm not sure if a number and a switch is the best way to represent it. I tagged it for second opinion wanted.

barneyonline committed:
Merge branch 'home-assistant:dev' into dev

barneyonline committed:
Switch from unittest to pytest

For async support

elupus commented:
I would also say a separate climate entity. It could be very limited in features. Just set point and on/off mode. It also support setting the +2/-2 range so i see no point in not using it.

Automating on it is no different than number entity really.

fredrike commented:
I'm pretty sure that not all devices that supports zones supports zone-temperature. Please show which device models will get this support.

barneyonline commented:
> I'm pretty sure that not all devices that supports zones supports zone-temperature. Please show which device models will get this support.

Correct, only compatible controllers and inverters that support linear zone control.
This implementation will only expose the zone temperature controls when they are supported.

From what I can find [here](https://www.daikin.com.au/controllers/daikin-airhub), a user needs:
- Daikin AirHub controller, and
- Linear Zone Controller Box, and
- R32 (FDYA) Premium Inverter Ducted 71-160 Class, or R32 (FDYAN) Inverter Ducted 50-160 Class

Where do you want this shown?

barneyonline commented:
@joostlek thinking a bit more about the number v climate entity approach, when I consider how a climate entity would be implemented here:

1. Mode (including on/off) cannot be set, but should probably reflect what the "main" controller is set to. Instead the on/off would need to be controlled through the separate zone switch. Can we display the mode as on and off even when the "main" controller is set to a specific mode (e.g. Heat)?
2. Current temperature is unknown for the zone, so assume we show the "main" controller temperature here. Looking through the pydaikin source it isn't possible to get the zone-specific current temperature, although the controller does know it.
3. Fan mode cannot be set, as the system uses "OptiZone" that auto-adjusts airflow. Assume we just show what is reported and ignore any changes (as per current implementation)
4. History will be a copy of the "main" controller but with a potentially different set temperature, which may be confusing when zone temperature is different to "main" temperature.

Implementing the climate entity is probably possible but I think it will be very confusing to users who don't understand the nuances of how the climate system works.

emontnemery commented:
Regarding the number vs climate entity, I agree a climate entity makes sense. Frontend adapts fairly well to an entity which only supports setting the target temperature, and nothing else:
![image](https://github.com/user-attachments/assets/a79bac3c-f4e5-472e-b49c-3c858cbf0a48)

Buttons for turning on or off, setting mode, setting a range, fan speed etc. are all gone.
Maybe it could be improved to not show the current temperature instead of "Unknown" though.

fredrike commented:
Wouldn't it be best to implement this as a service call and then the users can implement r whatever they want in the frontend.

barneyonline committed:
Merge branch 'home-assistant:dev' into daikin_zone_temperature_support

fredrike commented:
@barneyonline did you abonon this?

barneyonline commented:
@fredrike Afraid so. I've had a few attempts to redo this but havent been able to get it working on my personal system in the limited time I have.
